base/test.jl

```diff
@@ -782,6 +782,49 @@ macro test_approx_eq(a, b)
     :(test_approx_eq($(esc(a)), $(esc(b)), $(string(a)), $(string(b))))
 end
 
+"""
+    @inferred f(x)
+
+Tests that the call expression `f(x)` returns a value of the same type
+inferred by the compiler. It's useful to check for type stability.
+
+`f(x)` can be any call expression, except a call for a generated function.
+Returns the result of `f(x)` if the types match, and an `Error` `Result` if
+it finds different types.
+
+```jldoctest
+julia> using Base.Test
+
+julia> f(a,b,c) = b > 1 ? 1 : 1.0
+f (generic function with 1 method)
+
+julia> typeof(f(1,2,3))
+Int64
+
+julia> @code_warntype f(1,2,3)
+Variables:
+  #self#::#f
+  a::Int64
+  b::Int64
+  c::Int64
+
+Body:
+  begin  # REPL[2], line 1:
```

Comment:
see 632e0873574a04b3b3e734f53feec53814826d31, not sure how much of that is accurate for master too